### PR TITLE
Add Quest Bookmarking

### PR DESCRIPTION
**Pull request by Caedis**
image

Preview in #mod-dev channel:

https://discord.com/channels/181078474394566657/603348502637969419/1192604755952947291

### CODE REVIEW

File: src/main/java/betterquesting/questing/QuestInstance.java [hunk 1]
```diff
@@ -649,6 +670,34 @@ public void readProgressFromNBT(NBTTagCompound json, boolean merge)
...
 
+    @Override
+    public NBTTagCompound writePinnedToNBT(NBTTagCompound json, @Nullable List<UUID> users)
+    {
+        synchronized(pinnedUsers)
+        {
+            NBTTagList list = new NBTTagList();
+            for(UUID entry : pinnedUsers)
+            {
+                if(entry == null) continue;
```
Comment (Glease) [inline src/main/java/betterquesting/questing/QuestInstance.java:681]: how can it be null?

**Review by Glease — COMMENTED** [review]
1. save load is a bit awkward. most of write logic is in QuestInstance but most of read logic is in QuestDatabase. perhaps just move them all to QuestDatabase?
2. as mentioned on discord. this should probably be called favorite bookmark rather than pin. 
3. GUI code is always if it works then ship it. we don't check code quality in this area.

File: src/main/java/betterquesting/questing/QuestInstance.java (hunk 1 above)
Comment (Caedis) [reply src/main/java/betterquesting/questing/QuestInstance.java:681]: Was from the code I copied from

**Comment by Caedis** [comment]
Now the question is, what icon to make to use on the quest instead of the pins

image

**Comment by Caedis** [comment]
Took over the bookmark icon and change the quest line icon to a better match
image

**Comment by Caedis** [comment]
Hmm, losing bookmarks on a server

**Comment by Caedis** [comment]
Current issues:

- [ ] Bookmarks are not saving when on a server. This is because the quests are handled on the server.

**Comment by Lyfts** [comment]
Bookmarks are now handled by the client and are unique for each world/server.

**Comment by Caedis** [comment]
Looks good, i'll take a look at it in game later today

**Review by Caedis — COMMENTED** [review]
The order the quests are bookmarked should be kept. 

In the screenshot, the IV circuit was the last one I bookmarked.

image

**Comment by Caedis** [comment]
Also if a server and a sp world have the same name, the bookmark files will be shared i believe

**Comment by Caedis** [comment]
Cool, didnt even realize that colon were valid in file names

**Comment by eigenraven** [comment]
Colons are not valid in filenames on Windows: https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions

**Comment by Caedis** [comment]
Thats what i thought. Worked fine for me on linux

**Comment by Caedis** [comment]
maybe change it to ip-port

**Comment by Lyfts** [comment]
I'm already replacing the colon with a dot

**Comment by Caedis** [comment]
Oh, didnt notice that. Work works then

**Comment by Lyfts** [comment]
Alright bookmarks will now keep their order and I also implemented a simple method to skip iterating over the entire quest list if no bookmarks have changed.

**Comment by Caedis** [comment]
Perfect, thanks!
image

...i can't approve it since its my pr lol

**Comment by Caedis** [comment]
hmmm

**Comment by Caedis** [comment]
Something is not updating when I change SP worlds
Currently in a sp world with this bookmarked
image

but, the list is showing the bookmarks from the other world
image

**Comment by Caedis** [comment]
I had 2 bookmarks that were shared between the 2 worlds

**Comment by Lyfts** [comment]
Oh my bad I forgot to reset the flag when loading bookmarks

**Comment by Lyfts** [comment]
It seems to load correctly now, good catch.

**Comment by Caedis** [comment]
Perfect

**Review by Dream-Master — APPROVED** [review]
tested works

**Pull request merged**